desk/app/activity.hoon

```diff
@@ -108,6 +108,123 @@
     [%2 %all +.old]
   --
 ::
+++  correct-dm-keys
+  |^  ^+  cor
+  =+  .^  [dms=(map ship dm:ch) clubs=(map id:club:ch club:ch)]
+    %gx  (scry-path %chat /full/noun)
+  ==
+  =/  club-threads
+    %~  tap  by
+    %+  roll
+      ~(tap by indices)
+    |=  [[=source:a =index:a] dms=(jar whom:a [source:a index:a])]
+    ?.  ?=(?(%dm %dm-thread) -.source)  dms
+    =/  whom  ?-(-.source %dm whom.source, %dm-thread whom.source)
+    ?.  ?=(%club -.whom)  dms
+    (~(add ja dms) whom [source index])
+  |-
+  ?~  club-threads  cor
+  =/  [=whom:a =indexes]  -.club-threads
+  =*  next  $(club-threads +.club-threads)
+  ?.  ?=(%club -.whom)  next
+  =/  club  (~(get by clubs) p.whom)
+  ?~  club  next
+  =/  [threads=^indexes dms=^indexes]
+    %+  skid
+      indexes
+    |=  [=source:a *]
+    ?=(%dm-thread -.source)
+  =;  indxs=^indexes
+    |-
+    ?~  indxs  next
+    =*  source  source.i.indxs
+    ::  cleanup old bad keys
+    =?  cor  ?=(%dm-thread -.source)
+      =/  old-source  source(key [id.key.source q.id.key.source])
+      %=  cor
+        indices  (~(del by indices) old-source)
+        activity  (~(del by activity) old-source)
+        volume-settings  (~(del by volume-settings) old-source)
+      ==
+    ::  update source + index, if new key create new index
+    =.  cor  (update-index source index.i.indxs &)
```

Comment:
We do processing on the `%dm-thread` sources here exclusively, because those sources (source ids, indices, w/e) themselves have the incorrect keys, right?

Also, don't we want to do the `+update-index` call only once at the end, in the `?~ indxs` case?

Reply:
yes both the sources and the indexes need correction. and no we have to call update-index for each one since they are disparate indexes and each one needs to update.